fix the false 'defined here' messages

henryboisdequin opened the pull request:
Closes #80853.

Take this code:

```rust
struct S;

fn repro_ref(thing: S) {
    thing();
}
```

Previously, the error message would be this:

```
error[E0618]: expected function, found `S`
 --> src/lib.rs:4:5
  |
3 | fn repro_ref(thing: S) {
  |              ----- `S` defined here
4 |     thing();
  |     ^^^^^--
  |     |
  |     call expression requires function

error: aborting due to previous error
```

This is incorrect as `S` is not defined in the function arguments, `thing` is defined there. With this change, the following is emitted:

```
error[E0618]: expected function, found `S`
  --> $DIR/80853.rs:4:5
   |
LL | fn repro_ref(thing: S) {
   |              ----- is of type `S`
LL |     thing();
   |     ^^^^^--
   |     |
   |     call expression requires function
   |
   = note: local variable `S` is not a function

error: aborting due to previous error
```

As you can see, this error message points out that `thing` is of type `S` and later in a note, that `S` is not a function. This change does seem like a downside for some error messages. Take this example:

```
LL | struct Empty2;
   | -------------- is of type `Empty2`
```

As you can see, the error message shows that the definition of `Empty2` is of type `Empty2`. Although this isn't wrong, it would be more helpful if it would say something like this (which was there previously):

```
LL | struct Empty2;
   | -------------- `Empty2` defined here
```

If there is a better way of doing this, where the `Empty2` example would stay the same as without this change, please inform me.

**Update: This is now fixed**

CC @camelid

rust-highfive commented:
r? @varkor

(rust-highfive has picked a reviewer for you, use r? to override)

pickfire commented on src/test/ui/structs/80853.stderr:
```diff
@@ -0,0 +1,15 @@
+   |     |
+   |     call expression requires function
+   |
+   = note: local variable `S` is not a function
```
This is confusing for me, `S` is not a "local variable", isn't `S` a "unit struct" but it was being used as a generic?

pickfire reviewed (commented):
I don't know if adding the note is more helpful since it felt redundant. The error at the top already says so.
```
expected function, found `Empty2`
```

henryboisdequin commented:
> I don't know if adding the note is more helpful since it felt redundant. The error at the top already says so.
> 
> ```
> expected function, found `Empty2`
> ```
> 
> And we show
> 
> ```
> note: unit struct `Empty2` is not a function
> ```
> 
> If it errors with "expected function" means `Empty2` is not a function right? Why do we need to repeat?

Thanks for the review @pickfire. I agree, I'll remove that redundant note. By doing so, the local variable mistake is also fixed.

henryboisdequin commented:
@varkor This PR is now ready for review.

varkor commented on compiler/rustc_typeck/src/check/callee.rs:
```diff
@@ -374,7 +374,16 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                                // If we're dealing with a local variable, its not
+                                // the struct which is defined there but the
+                                // variable is of type some struct
+                                if def.descr() == "local variable" {
```
It would be better to check directly that you have a `Res::Local`.

varkor commented on compiler/rustc_typeck/src/check/callee.rs:
```diff
@@ -374,7 +374,16 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                                // the struct which is defined there but the
+                                // variable is of type some struct
+                                if def.descr() == "local variable" {
+                                    Some(format!("is of type `{}`", callee_ty))
```
Can we display the variable name too? E.g.
```
`foo` has type `u32`
```

henryboisdequin replied:
I think that this is the whole point of the PR. In this case (where it's a local variable), `callee_ty` is the type of variable. Right now, in the case where we are dealing with a local variable, I don't have access to the variable name which has a type of `callee_ty`. Please correct me if I'm wrong.

varkor replied:
When `def` is a `Res::Local`, we know it refers to the path of a variable. I would imagine it's possible to extract the name in that case (I don't have time to check right now, so let me know if you can't figure out how to do it).

henryboisdequin replied:
@varkor Thanks for all your comments, I've addressed 2 out of the 3. I'll work on this one.

pickfire replied:
```suggestion
                                    Some(format!("has type `{}`", callee_ty))
```

IMHO "has type" is clearer than "is of type"

varkor commented on compiler/rustc_typeck/src/check/callee.rs:
```diff
@@ -374,7 +374,16 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                             }
-                            _ => Some(format!("`{}` defined here", callee_ty)),
+                            _ => {
+                                // If we're dealing with a local variable, its not
+                                // the struct which is defined there but the
+                                // variable is of type some struct
```
```suggestion
                                // If `def` is a local variable, it means that it is not a
                                // definition itself, but a variable of that type, which means we
                                // need to emit a different diagnostic.
```

henryboisdequin commented:
@rustbot label +S-waiting-on-author -S-waiting-on-review

henryboisdequin commented:
@rustbot label -S-waiting-on-author +S-waiting-on-review

pickfire commented on compiler/rustc_typeck/src/check/callee.rs:
```diff
@@ -374,7 +374,20 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                             }
-                            _ => Some(format!("`{}` defined here", callee_ty)),
+                            _ => {
+                                // If `def` is a local variable, it means that it is not a
+                                // definition itself, but a variable of that type, which means we
+                                // need to emit a different diagnostic.
```
```suggestion
                                // definition itself but a variable of that type.
```

The action part is obvious, the explanation is good enough, I was confused reading the action.

pickfire replied:
```suggestion
                                // Emit a different diagnostic for local variable as it is
                                // not a definition itself but a variable of that type.
```

Maybe this would be better? I find the last part a bit hard to read (need to read a few times).

varkor commented on compiler/rustc_typeck/src/check/callee.rs:
```diff
@@ -378,8 +378,12 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
-                                    Some(format!("is of type `{}`", callee_ty))
+                                if let Res::Local(id) = def {
+                                    Some(format!(
+                                        "{} has type `{}`",
```
```suggestion
                                        "`{}` has type `{}`",
```

varkor commented on compiler/rustc_typeck/src/check/callee.rs:
```diff
@@ -374,7 +374,19 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                             }
-                            _ => Some(format!("`{}` defined here", callee_ty)),
+                            _ => {
+                                // Emit a different diagnostic for local variable as it is
+                                // not a definition itself but a variable of that type.
```
```suggestion
                                // Emit a different diagnostic for local variables, as they are not
                                // type definitions themselves, but rather variables *of* that type.
```

varkor commented:
If you could address the last two comments and squash the commits, I think this is good to go!

camelid commented:
There is another case that I don't think this handles:

```rust
const FOO: usize = 0;

fn main() {
    FOO();
}
```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=bed98b8f5347f4cfd9cd0ea08ea71cc9))

Errors:

```
   Compiling playground v0.0.1 (/playground)
error[E0618]: expected function, found `usize`
 --> src/main.rs:4:5
  |
1 | const FOO: usize = 0;
  | --------------------- `usize` defined here
...
4 |     FOO();
  |     ^^^--
  |     |
  |     call expression requires function

error: aborting due to previous error

For more information about this error, try `rustc --explain E0618`.
error: could not compile `playground`

To learn more, run the command again with --verbose.
```

camelid commented:
Perhaps the fix would be also checking for `Res::Def` where `DefKind::ns` is `ValueNS`?

henryboisdequin commented:
@rustbot label +S-waiting-on-author -S-waiting-on-review

henryboisdequin commented:
Sorry for taking a while, I was on vacation. @varkor, I believed I have addressed all your comments. @camelid I did the check that you have suggested but it does bring us back to the original problem, like this:

```
LL | struct Empty2;
   | -------------- is of type `Empty2`
```

It should say `Empty2 defined here`. Is there a better way to do this? Also, is there a way to get the name of a `const` to be able to say `FOO has type usize` instead of `-- has type usize`. Thanks

camelid commented on compiler/rustc_typeck/src/check/callee.rs:
```diff
@@ -374,7 +374,25 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                                        self.tcx.hir().name(id),
+                                        callee_ty
+                                    )),
+                                    Res::Def(kind, _) => {
+                                        if let Some(Namespace::ValueNS) = kind.ns() {
+                                            Some(format!("has type `{}`", callee_ty))
```
This should show the name:

```suggestion
                                    Res::Def(kind, def_id) => {
                                        if let Some(Namespace::ValueNS) = kind.ns() {
                                            Some(format!("`{}` has type `{}`", self.tcx.def_path_str(def_id), callee_ty))
```

henryboisdequin commented on src/test/ui/issues/issue-21701.stderr:
```diff
@@ -12,7 +12,7 @@ error[E0618]: expected function, found `Bar`
    |
 LL | struct Bar;
-   | ----------- `Bar` defined here
+   | ----------- `Bar` has type `Bar`
```
A little sad...

camelid commented:
Not sure if it's worth adding a cross-crate test (where `const FOO` is in one crate and it's called in another crate) to test `def_path_str`'s behavior, but not a big deal either way.

Thanks for implementing this! The errors look much better now.

camelid commented on compiler/rustc_typeck/src/check/callee.rs:
```diff
@@ -374,7 +374,29 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                                    )),
+                                    Res::Def(kind, def_id) => {
+                                        if let Some(Namespace::ValueNS) = kind.ns() {
+                                            Some(format!(
+                                                "`{}` has type `{}`",
+                                                self.tcx.def_path_str(def_id),
+                                                callee_ty
+                                            ))
```
> @camelid I did the check that you have suggested but it does bring us back to the original problem, like this:
> 
> ```
> LL | struct Empty2;
>    | -------------- is of type `Empty2`
> ```
> 
> It should say `Empty2 defined here`. Is there a better way to do this?

I think it's fine the way it is now—in fact, I think it's better than the `Empty2 defined here` that I originally suggested—but if you don't want to show the type, I think you can just do

```suggestion
                                            Some(format!(
                                                "`{}` defined here",
                                                self.tcx.def_path_str(def_id),
                                            ))
```

henryboisdequin commented:
@camelid I've addressed your comments. Thanks for pointing me in the right direction

henryboisdequin commented:
@rustbot label -S-waiting-on-author +S-waiting-on-review

camelid commented:
> @camelid I've addressed your comments. Thanks for pointing me in the right direction

Oh, I actually thought the errors were better before; I was just suggesting that so you knew how :)

But, it's not a big deal either way.

pickfire commented on compiler/rustc_typeck/src/check/callee.rs:
```diff
@@ -374,7 +374,28 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                                        self.tcx.hir().name(hir_id),
+                                        callee_ty
+                                    )),
+                                    Res::Def(kind, def_id) => {
+                                        if let Some(Namespace::ValueNS) = kind.ns() {
+                                            Some(format!(
+                                                "`{}` defined here",
+                                                self.tcx.def_path_str(def_id),
+                                            ))
+                                        } else {
+                                            Some(format!("`{}` defined here", callee_ty))
+                                        }
+                                    }
```
```suggestion
                                    Res::Def(kind, def_id) if let Some(Namespace::ValueNS) = kind.ns() => {
                                        Some(format!(
                                            "`{}` defined here",
                                            self.tcx.def_path_str(def_id),
                                        ))
                                    }
```
Since this is just a one time thing we can probably simplify it with an if let match guard.

pickfire replied:
Why was this resolved?

henryboisdequin replied:
@pickfire Do you know why this doesn't work?

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1958521657) [(plain)](https://github.com/rust-lang/rust/commit/4e2a8335a352aef52a585fb4206e011abeb28acd/checks/1958521657/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_mir_build v0.0.0 (/checkout/compiler/rustc_mir_build)
    Checking rustc_mir v0.0.0 (/checkout/compiler/rustc_mir)
    Checking rustc_traits v0.0.0 (/checkout/compiler/rustc_traits)
    Checking rustc_passes v0.0.0 (/checkout/compiler/rustc_passes)
error[E0658]: `if let` guards are experimental
   --> compiler/rustc_typeck/src/check/callee.rs:386:60
    |
386 | ...                   Res::Def(kind, def_id) if let Some(Namespace::ValueNS) = kind.ns() => {
    |
    = note: see issue #51114 <https://github.com/rust-lang/rust/issues/51114> for more information
    = note: see issue #51114 <https://github.com/rust-lang/rust/issues/51114> for more information
    = help: add `#![feature(if_let_guard)]` to the crate attributes to enable
    Checking rustc_plugin_impl v0.0.0 (/checkout/compiler/rustc_plugin_impl)
error: aborting due to previous error

For more information about this error, try `rustc --explain E0658`.
For more information about this error, try `rustc --explain E0658`.
error: could not compile `rustc_typeck`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--color" "always" "--features" " llvm max_level_info" "--manifest-path" "/checkout/compiler/rustc/Cargo.toml" "-p" "rustc-main" "-p" "rustc_driver" "-p" "rustc_hir" "-p" "rustc_macros" "-p" "rustc_index" "-p" "rustc_feature" "-p" "rustc_ast" "-p" "rustc_lexer" "-p" "rustc_serialize" "-p" "rustc_middle" "-p" "rustc_type_ir" "-p" "rustc_apfloat" "-p" "rustc_arena" "-p" "rustc_attr" "-p" "rustc_query_system" "-p" "rustc_error_codes" "-p" "rustc_plugin_impl" "-p" "rustc_target" "-p" "rustc_mir" "-p" "coverage_test_macros" "-p" "rustc_graphviz" "-p" "rustc_trait_selection" "-p" "rustc_parse_format" "-p" "rustc_infer" "-p" "rustc_save_analysis" "-p" "rustc_data_structures" "-p" "rustc_interface" "-p" "rustc_traits" "-p" "rustc_builtin_macros" "-p" "rustc_expand" "-p" "rustc_lint_defs" "-p" "rustc_ty_utils" "-p" "rustc_privacy" "-p" "rustc_symbol_mangling" "-p" "rustc_incremental" "-p" "rustc_fs_util" "-p" "rustc_ast_lowering" "-p" "rustc_passes" "-p" "rustc_typeck" "-p" "rustc_ast_passes" "-p" "rustc_codegen_llvm" "-p" "rustc_llvm" "-p" "rustc_query_impl" "-p" "rustc_mir_build" "-p" "rustc_resolve" "-p" "rustc_parse" "-p" "rustc_span" "-p" "rustc_session" "-p" "rustc_metadata" "-p" "rustc_lint" "-p" "rustc_hir_pretty" "-p" "rustc_ast_pretty" "-p" "rustc_errors" "-p" "rustc_codegen_ssa" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check
Build completed unsuccessfully in 0:02:21
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1960577468) [(plain)](https://github.com/rust-lang/rust/commit/2e18f199b1f714a6fcef6d20f3b40ce258b58f30/checks/1960577468/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_ty_utils v0.0.0 (/checkout/compiler/rustc_ty_utils)
    Checking rustc_mir_build v0.0.0 (/checkout/compiler/rustc_mir_build)
    Checking rustc_passes v0.0.0 (/checkout/compiler/rustc_passes)
    Checking rustc_mir v0.0.0 (/checkout/compiler/rustc_mir)
error[E0658]: `if let` guards are experimental
   --> compiler/rustc_typeck/src/check/callee.rs:388:60
    |
388 | ...                   Res::Def(kind, def_id) if let Some(Namespace::ValueNS) = kind.ns() => {
    |
    = note: see issue #51114 <https://github.com/rust-lang/rust/issues/51114> for more information
    = note: see issue #51114 <https://github.com/rust-lang/rust/issues/51114> for more information
    = help: add `#![feature(if_let_guard)]` to the crate attributes to enable
    Checking rustc_plugin_impl v0.0.0 (/checkout/compiler/rustc_plugin_impl)
error: aborting due to previous error

For more information about this error, try `rustc --explain E0658`.
For more information about this error, try `rustc --explain E0658`.
error: could not compile `rustc_typeck`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--color" "always" "--features" " llvm max_level_info" "--manifest-path" "/checkout/compiler/rustc/Cargo.toml" "-p" "rustc-main" "-p" "rustc_driver" "-p" "rustc_data_structures" "-p" "rustc_graphviz" "-p" "rustc_macros" "-p" "rustc_index" "-p" "rustc_middle" "-p" "rustc_arena" "-p" "rustc_attr" "-p" "rustc_lexer" "-p" "rustc_apfloat" "-p" "rustc_query_system" "-p" "rustc_type_ir" "-p" "rustc_hir" "-p" "rustc_plugin_impl" "-p" "rustc_hir_pretty" "-p" "rustc_metadata" "-p" "rustc_expand" "-p" "rustc_lint_defs" "-p" "rustc_ast_passes" "-p" "rustc_error_codes" "-p" "rustc_ast" "-p" "rustc_interface" "-p" "rustc_resolve" "-p" "rustc_builtin_macros" "-p" "rustc_parse_format" "-p" "rustc_incremental" "-p" "rustc_fs_util" "-p" "rustc_typeck" "-p" "rustc_infer" "-p" "rustc_ty_utils" "-p" "rustc_privacy" "-p" "rustc_mir_build" "-p" "rustc_passes" "-p" "rustc_ast_lowering" "-p" "rustc_codegen_llvm" "-p" "rustc_llvm" "-p" "rustc_symbol_mangling" "-p" "rustc_query_impl" "-p" "rustc_traits" "-p" "rustc_trait_selection" "-p" "rustc_session" "-p" "rustc_serialize" "-p" "rustc_feature" "-p" "rustc_mir" "-p" "coverage_test_macros" "-p" "rustc_ast_pretty" "-p" "rustc_errors" "-p" "rustc_save_analysis" "-p" "rustc_lint" "-p" "rustc_target" "-p" "rustc_span" "-p" "rustc_parse" "-p" "rustc_codegen_ssa" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check
Build completed unsuccessfully in 0:02:35
```

</details>

pickfire commented on compiler/rustc_typeck/src/check/callee.rs:
```diff
@@ -1,10 +1,12 @@
+#![feature(if_let_guard)]
```
I believe you need to add it to `compiler/rustc_typeck/src/lib.rs` rather than here.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1961275819) [(plain)](https://github.com/rust-lang/rust/commit/01fbc5b4e423eae87d8dcc056a8a26b9ac46259e/checks/1961275819/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Removing intermediate container 587a53b3e661
 ---> f751d51eb8ba
Step 5/10 : RUN npm install es-check -g
 ---> Running in 403921a076e5
/node-v14.4.0-linux-x64/bin/es-check -> /node-v14.4.0-linux-x64/lib/node_modules/es-check/index.js

> spawn-sync@1.0.15 postinstall /node-v14.4.0-linux-x64/lib/node_modules/es-check/node_modules/spawn-sync
> node postinstall
+ es-check@5.2.0
added 95 packages from 44 contributors in 4.572s
Removing intermediate container 403921a076e5
 ---> 85f1df963e8a
---
 ---> b97cbbeecb5f
Successfully built b97cbbeecb5f
Successfully tagged rust-ci:latest
Built container sha256:b97cbbeecb5ff54b8ff7859c3f4013e301988f8321c8c6c0fa657dd72d3260d9
Uploading finished image to https://ci-caches.rust-lang.org/docker/9f2a38e35a8211f9c2c342213b6dabbf1ce1e957c3f9f4a6874af054aa93d446d1c6f8252277cb4118d76ddf7862eec7c972b385df9acf97d8b518b20c0181e6
upload failed: - to s3://rust-lang-ci-sccache2/docker/9f2a38e35a8211f9c2c342213b6dabbf1ce1e957c3f9f4a6874af054aa93d446d1c6f8252277cb4118d76ddf7862eec7c972b385df9acf97d8b518b20c0181e6 Unable to locate credentials
[CI_JOB_NAME=mingw-check]
---
    Checking rustc_mir v0.0.0 (/checkout/compiler/rustc_mir)
    Checking rustc_ty_utils v0.0.0 (/checkout/compiler/rustc_ty_utils)
    Checking rustc_mir_build v0.0.0 (/checkout/compiler/rustc_mir_build)
    Checking rustc_traits v0.0.0 (/checkout/compiler/rustc_traits)
error: the feature `if_let_guard` is incomplete and may not be safe to use and/or cause compiler crashes
   |
   |
72 | #![feature(if_let_guard)]
   |
   |
   = note: `-D incomplete-features` implied by `-D warnings`

    Checking rustc_plugin_impl v0.0.0 (/checkout/compiler/rustc_plugin_impl)
error: aborting due to previous error


error: could not compile `rustc_typeck`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--color" "always" "--features" " llvm max_level_info" "--manifest-path" "/checkout/compiler/rustc/Cargo.toml" "-p" "rustc-main" "-p" "rustc_driver" "-p" "rustc_plugin_impl" "-p" "rustc_metadata" "-p" "rustc_expand" "-p" "rustc_ast_passes" "-p" "rustc_lint_defs" "-p" "rustc_lexer" "-p" "rustc_macros" "-p" "rustc_attr" "-p" "rustc_index" "-p" "rustc_ast" "-p" "rustc_middle" "-p" "rustc_apfloat" "-p" "rustc_arena" "-p" "rustc_type_ir" "-p" "rustc_query_system" "-p" "rustc_target" "-p" "rustc_save_analysis" "-p" "rustc_lint" "-p" "rustc_trait_selection" "-p" "rustc_infer" "-p" "rustc_graphviz" "-p" "rustc_parse_format" "-p" "rustc_mir" "-p" "coverage_test_macros" "-p" "rustc_interface" "-p" "rustc_passes" "-p" "rustc_query_impl" "-p" "rustc_privacy" "-p" "rustc_mir_build" "-p" "rustc_builtin_macros" "-p" "rustc_ty_utils" "-p" "rustc_typeck" "-p" "rustc_resolve" "-p" "rustc_incremental" "-p" "rustc_fs_util" "-p" "rustc_codegen_llvm" "-p" "rustc_llvm" "-p" "rustc_symbol_mangling" "-p" "rustc_ast_lowering" "-p" "rustc_traits" "-p" "rustc_parse" "-p" "rustc_errors" "-p" "rustc_hir" "-p" "rustc_hir_pretty" "-p" "rustc_feature" "-p" "rustc_error_codes" "-p" "rustc_ast_pretty" "-p" "rustc_session" "-p" "rustc_data_structures" "-p" "rustc_serialize" "-p" "rustc_span" "-p" "rustc_codegen_ssa" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check
Build completed unsuccessfully in 0:02:48
```

</details>

henryboisdequin commented:
> The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1961275819) [(plain)](https://github.com/rust-lang/rust/commit/01fbc5b4e423eae87d8dcc056a8a26b9ac46259e/checks/1961275819/logs)
> Click to see the possible cause of the failure (guessed by this bot)

@pickfire, should I just change it to where it was before

pickfire commented on compiler/rustc_typeck/src/lib.rs:
```diff
@@ -68,6 +68,7 @@ This API is completely unstable and subject to change.
 #![feature(never_type)]
 #![feature(slice_partition_dedup)]
 #![feature(control_flow_enum)]
+#![feature(if_let_guard)]
```
```suggestion
#![feature(if_let_guard)]
#![allow(incomplete_features)]
```

Maybe try this?

pickfire reviewed (approved):
Looks good to me.

varkor commented:
This looks great @henryboisdequin, thanks! Could you squash your commits? Then it's good to go.

henryboisdequin commented:
@varkor Commits are squashed ✅

varkor commented:
@henryboisdequin: there are some commits that don't appear to be squashed. It's fine to separate different changes into separate commits, but when later commits are reverting changes in previous commits, it just creates churn (which makes tools like `git blame` less convenient). This change is small enough that it's readable as only one commit.

henryboisdequin commented:
@varkor Done

varkor commented:
@henryboisdequin: thanks for improving this diagnostic!

@bors r+

bors commented:
:pushpin: Commit 7d85592713c495e4b746c06e78bd8f90ab6653cd has been approved by `varkor`

<!-- @bors r=varkor 7d85592713c495e4b746c06e78bd8f90ab6653cd -->
<!-- homu: {"type":"Approved","sha":"7d85592713c495e4b746c06e78bd8f90ab6653cd","approver":"varkor"} -->

camelid commented on compiler/rustc_typeck/src/lib.rs:
```diff
@@ -68,6 +68,8 @@ This API is completely unstable and subject to change.
 #![feature(slice_partition_dedup)]
 #![feature(control_flow_enum)]
+#![feature(if_let_guard)]
+#![allow(incomplete_features)]
```
Is it okay to use incomplete features in rustc? Based on a quick ripgrep, I can't find any other instances of `allow(incomplete_features)`.

henryboisdequin replied:
Should I remove it?

camelid replied:
Yes, could you please remove the use of the `if_let_guard` feature? @bors r-

camelid commented on compiler/rustc_typeck/src/check/callee.rs:
```diff
@@ -374,7 +374,24 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                                        self.tcx.hir().name(hir_id),
+                                        callee_ty
+                                    )),
+                                    Res::Def(kind, def_id) if let Some(Namespace::ValueNS) = kind.ns() => {
```
I think this can just be changed to:

```suggestion
                                    Res::Def(kind, def_id) if kind.ns() == Some(Namespace::ValueNS) => {
```

(with the feature gate and `allow(incomplete_features)` removed as well)

henryboisdequin commented:
@varkor, @camelid Done ✅

varkor commented:
@camelid: thanks for pointing that out.

@bors r+

bors commented:
:pushpin: Commit d7cb66d389c3ef542891a26ee3c12308b407fcd6 has been approved by `varkor`

<!-- @bors r=varkor d7cb66d389c3ef542891a26ee3c12308b407fcd6 -->
<!-- homu: {"type":"Approved","sha":"d7cb66d389c3ef542891a26ee3c12308b407fcd6","approver":"varkor"} -->
